<?xml version="1.0"?>

<launch>

    <!-- launch the zed_ros_wrapper node-->
    <include file="$(find zed_wrapper)/launch/zed2.launch">
    </include>

    <node name="orb_slam2_stereo" pkg="orb_slam2_ros"
      type="orb_slam2_ros_stereo" output="screen">

       <remap from="image_left/image_color_rect" to="/zed2/zed_node/left/image_rect_color" />
       <remap from="image_right/image_color_rect" to="/zed2/zed_node/right/image_rect_color" />

       <param name="publish_pointcloud" type="bool" value="true" />
       <param name="publish_pose" type="bool" value="true" />
       <param name="localize_only" type="bool" value="false" />
       <param name="reset_map" type="bool" value="false" />

       <!-- static parameters -->
       <param name="load_map" type="bool" value="false" />
       <param name="map_file" type="string" value="map.bin" />
       <param name="settings_file" type="string" value="$(find tx2_slam)/config/Zed2Stereo.yaml" />
       <param name="voc_file" type="string" value="$(find tx2_slam)/Vocabulary/ORBvoc.txt" />

       <param name="pointcloud_frame_id" type="string" value="map" />
       <param name="camera_frame_id" type="string" value="camera_link" />
       <param name="min_num_kf_in_map" type="int" value="5" />
    </node>

    <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_pos_x"             default="0.0" />
    <arg name="cam_pos_y"             default="0.0" />
    <arg name="cam_pos_z"             default="0.0" />

    <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="cam_roll"              default="0.0" />
    <arg name="cam_pitch"             default="0.0" />
    <arg name="cam_yaw"               default="0.0" />

    <!-- ROS URDF description of the ZED -->
    <group if="true">
        <param name="zed2_description"
               command="$(find xacro)/xacro '$(find tx2_slam)/urdf/zed_descr.urdf.xacro'
                        camera_name:=tx2zed2
                        camera_model:=zed2
                        base_frame:=zed2_base_link
                        cam_pos_x:=$(arg cam_pos_x)
                        cam_pos_y:=$(arg cam_pos_y)
                        cam_pos_z:=$(arg cam_pos_z)
                        cam_roll:=$(arg cam_roll)
                        cam_pitch:=$(arg cam_pitch)
                        cam_yaw:=$(arg cam_yaw)"
        />
        <node name="zed2_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen" required="true">
            <remap from="robot_description" to="zed2_description" />
        </node>
    </group>

    <!-- Launch rivz display -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find tx2_slam)/config/rviz_config.rviz" output="screen" />

    <node name="MapBuild" pkg="tx2_slam" type="MapBuild" output="screen"/>

</launch>
